### DISCUSSION TITLE
Link scanning software and recovery links

### DISCUSSION THREAD

**Discussion by kszafran** [ory/kratos#2563] [Q&A] [answered]
We run into an interesting problem with one of our customers. They seems to be using some kind of link scanning software (it looks like it might be https://campus.barracuda.com/product/essentials/doc/49055519/understanding-link-protection/).

When they receive the recovery email, the recovery link is transformed to something like: https://linkprotect.cudasvc.com/url?a=http%3A%2F%2Foriginal.url.here%26token=...&some-query-params.... And when they click on the link they land on the recovery page with an error saying that the link has expired or has already been used. My best guess is that this is what happens:
1. a recovery email is sent
2. the email is churned through their system that converts links
3. user clicks the link
4. the browser open the link, i.e. makes a call to https://linkprotect.cudasvc.com
5. the linkprotect backend calls our link (passed in the `a=` param) -> this "burns" the recovery link
6. linkprotect decides that the site is safe, so it redirects the user to the original URL
7. you end up with an error, because the link has already been used

The best course of action so far seems to talk to the customer's IT department to get our URL whitelisted. Does anyone have any thoughts on this? How do other user management systems deal with this? I was thinking of inspecting the User-Agent header and acting differently if it doesn't look like a browser (either in Kratos or on a reverse proxy). But I can easily imagine this or similar software simply using the browser's original User-Agent when calling Kratos, so this likely won't work in practice. Any other ideas?

**Comment by vinckr** [comment, marked as answer]
Hello @kszafran 
this is a common issue: https://github.com/ory/kratos/issues/1202, https://github.com/ory/kratos/issues/1142

We are working on TOTP codes which will be an alternative to magic links: https://github.com/ory/kratos/issues/1451

**Reply by kszafran** [reply]
Thanks for those issue links, somehow I didn't find them!
